File: vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>

int shift(char c);

int main(int argc, string argv[])
{
    //コマンドライン引数の個数が2つ以外の場合、強制終了
    if (argc != 2)
    {
        printf("Usage: ./vigenere keyword\n");
        return 1;
    }
    
    //コマンドライン引数の全ての文字がアルファベット以外の場合、強制終了
    for (int i = 0, len = strlen(argv[1]); i < len; i++)
    {
        if (!(isalpha(argv[1][i])))
        {
            printf("Usage: ./vigenere keyword\n");
            return 1;
        }
    }
    
    //暗号鍵の1文字目から順に配列shiftNumに格納する。
    int shiftNum[strlen(argv[1])];
    for (int i = 0, len = strlen(argv[1]); i < len ; i++)
    {
        shiftNum[i] = shift(argv[1][i]);
    }
    
    //平文の入力および表示
    printf("plaintext: ");
    string plainText = get_string("");
    
    //暗号文の表示
    printf("ciphertext: ");
    
    int key;
    int keyCount = 0;
    for (int i = 0, len = strlen(plainText); i < len; i++)
    {
        //暗号鍵を変数keyに格納(平文のi番目文字列÷暗号文字列の余剰)
        key =  shiftNum[keyCount % strlen(argv[1])];
        
        int asciiNum = plainText[i];
            
        //大文字の変換
        if ('A' <= asciiNum && asciiNum <= 'Z')
        {
            asciiNum = (asciiNum + key - 'A') % 26 + 'A';
            keyCount += 1;
        }
        
        //小文字の変換
        else if ('a' <= asciiNum && asciiNum <= 'z')
        {
            asciiNum = (asciiNum + key - 'a') % 26 + 'a';
            keyCount += 1;
        }
        printf("%c", asciiNum);
    }
    
    printf("\n");     
}

//文字をシフト
int shift(char c)
{
    int shiftNum = 0;
    if ('A' <= c && c <= 'Z')
    {
        shiftNum = (int)(c - 'A');
    }
    else if ('a' <= c && c <= 'z')
    {
        shiftNum = (int)(c - 'a');
    }
    return shiftNum;
}

File: Shinpei2-cs50-2019-x-sentimental-vigenere-20190725T174413Z/vigenere.py
from cs50 import get_string
import sys
argv = sys.argv

# If the number of command-line arguments isn't three , then finish the program.
if len(argv) != 2:
    print("Usage: python caesar.py k")
    sys.exit(1)


# 全ての文字が英字でない場合は強制終了
if argv[1].isalpha() is not True:
    print("Usage: python caesar.py k")
    sys.exit(1)

# input plaintext
plain = get_string("plaintext: ")
textLen = len(plain)

# keyの準備
key = 0
keyText = argv[1].upper()       # key文字列を大文字にして変数keyTextに格納
keyCount = 0
keyLen = len(keyText)
keyNum = 0

# create cipher text
cipherText = ""
for i in range(textLen):
    # keyの設定
    keyNum = keyCount % keyLen
    key = ord(keyText[keyNum]) - ord('A')

    # 変換前後のAscii値を変数に格納
    preAscii = ord(plain[i])        # ''で囲むと文字列と認識される点に注意
    newAscii = ord(plain[i])

    # 文字が大文字である場合の処理
    if ord('A') <= preAscii <= ord('Z'):
        newAscii = (preAscii + key - ord('A')) % 26 + ord('A')
        # print("uppercase")
        # keyCountを次に進める
        keyCount += 1

    # 文字が小文字である場合の処理
    if ord('a') <= preAscii <= ord('z'):
        newAscii = (preAscii + key - ord('a')) % 26 + ord('a')
        # print("lowercase")
        # keyCountを次に進める
        keyCount += 1

    # 変換処理後の文字をcipehrtextに追加
    cipherText += chr(newAscii)

# ciphertextを出力
print("ciphertext:", cipherText)
